presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -365,37 +358,70 @@ private QuerySpecification getRewrittenQuerySpecification(Metadata metadata, Qua
+            this.materializedViewName = requireNonNull(materializedViewName, "materialized view name is null");
         }
 
         public QuerySpecification rewrite(QuerySpecification querySpecification)
         {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            if (materializedViewInfo.getErrorMessage().isPresent()) {
+                log.debug("Failed to rewrite query with materialized view with following exception: %s", materializedViewInfo.getErrorMessage().get());
+                return querySpecification;
+            }
             // TODO: Implement ways to handle non-optimizable query without throw/catch. https://github.com/prestodb/presto/issues/16541
             try {
-                MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
-                materializedViewInformationExtractor.process(materializedViewQuery);
-                materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
-                return (QuerySpecification) process(querySpecification);
+                QuerySpecification rewrittenQuerySpecification = (QuerySpecification) process(querySpecification);
+
+                if (errorMessage.isPresent()) {
+                    log.debug("Rewrite optimization failed with: %s", errorMessage.get());
+                    return querySpecification;
+                }
+
+                if (rewrittenQuerySpecification == querySpecification) {
+                    return querySpecification;
+                }
+
+                if (!isMaterializedViewDataConsistencyEnabled(session)) {
+                    session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
+                    return rewrittenQuerySpecification;
+                }
+
+                // TODO: We should be able to leverage this information in the StatementAnalyzer as well.
+                MaterializedViewStatus materializedViewStatus = getMaterializedViewStatus(querySpecification);
+                if (materializedViewStatus.isPartiallyMaterialized() || materializedViewStatus.isFullyMaterialized()) {
+                    session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_SUBQUERY_COUNT, NONE, 1);
+                    return rewrittenQuerySpecification;
+                }
+                session.getRuntimeStats().addMetricValue(MANY_PARTITIONS_MISSING_IN_MATERIALIZED_VIEW_COUNT, NONE, 1);
+                return querySpecification;
             }
             catch (Exception e) {
+                log.info("Materialized view optimization failed exceptionally with: %s", e.getMessage());
```

Comment:
Use a consistent log level, we are using different level at different places.

Reply:
See https://github.com/prestodb/presto/pull/17911#discussion_r929428434